### DISCUSSION TITLE
[Concept] Anything Everywhere? Plus - A multi input UE? node

### DISCUSSION THREAD

**Discussion by LukeG89** [chrisgoringe/cg-use-everywhere#165] [Ideas]
@chrisgoringe @JorgeR81 Yesterday when we were talking about merging multiple UE? nodes to make one single node, I had this idea: A new UE? node (temporary called _Anything Everywhere? Plus_) that can take multiple inputs at once.

This is a mockup of the node:
![Plus_mockup_concept](https://github.com/chrisgoringe/cg-use-everywhere/assets/161304036/08f9dda9-bcba-49f2-a46f-8d26e78335d8)

I took this concept from other nodes that adds inputs, like this switch:
![connection_example](https://github.com/chrisgoringe/cg-use-everywhere/assets/161304036/a267f15b-9109-465c-b930-78cf569cd177)

I don't know if it's possible, just wanted to drop this idea here.

What do you think?

**Comment by JorgeR81** [comment]
That's an interesting idea. 

But I think it would be more convenient to have a version of this with a single [ title_regex ] and [ group_regex ] for all UE? connections. 

If it's possible to make the [ Simple String ] node to work in grouped nodes ( https://github.com/chrisgoringe/cg-use-everywhere/issues/160 ) we could use grouping to create something similar. 

But, the ability to add more connections, "on the fly", would be great !

![g1](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/2df00383-332c-4d68-8c5f-7330bfd6156a)

![g11](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/7209e0fe-7ad1-47b4-bbd5-cffa1527465a)

**Comment by LukeG89** [comment]
> But I think it would be more convenient to have a version of this with a single [ title_regex ] and [ group_regex ] for all UE? connections.

Maybe those options can be integrated by adding a couple of switches to the node:

![same_regex_switches](https://github.com/chrisgoringe/cg-use-everywhere/assets/161304036/b1e9d77b-c9c0-49d6-87c1-61d9f65635a1)

**Comment by JorgeR81** [comment]
Yes, but if you have this set to [ false ], and then turn to [ true ], a lot of connections would change at the same time.

Maybe it would be easier to use with just 2 separate types of "plus" nodes, for each use case.

**Comment by LukeG89** [comment]
Just made another mockup with the switches:

![Plus_mockup_concept_with_switches](https://github.com/chrisgoringe/cg-use-everywhere/assets/161304036/0e3ce5f3-1f29-454d-a52b-11061557020b)

> Yes, but if you have this set to [ false ], and then turn to [ true ], a lot of connections would change at the same time.

Well, it all depends on what you need the node for: If you have to send everything to a certain node or group, you turn one or both the switches on; otherwise you keep all separate.

However, the funny part is that we are just speculating about something that we don't know if it can be made 😆

**Comment by JorgeR81** [comment]
Well, for the sake of speculation ...

Instead of switches we could have these options in the right click menu, like we have with group and color restrictions.

These options would not be changed very often, and this way the node could have a cleaner look. 

And using the right click menu, we could have other options, like "hide all input_regex ", like we can do with grouped nodes.

![rc](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/dfb8be42-403a-4c27-8c7a-cc69fd66cfc4)

![group](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/a7c37b98-49bc-43e4-83f9-bacac6f0d2c7)

**Comment by LukeG89** [comment]
That would be awesome! 🔥

**Comment by chrisgoringe** [comment]
Let me see if I've got this idea... this is the way I would imagine it working...

A brand new node - working name `Any Plus` which has the following characteristics:

- starts with a single input like the current vanilla UE node
- when something is connected, a new input is automatically created
- has a right-click option to expose the regex fields, which turns it into a `...?` node
  - I think one set of regexes for all the inputs, otherwise, it's a UI disaster. If you need different regexes, use another node.
  - If this is selected, the regexes are edited, and then it is unselected, does it remain in `...?` mode? ie do hidden regexes apply? My instinct is to say no - you are toggling regex-based restrictions (like you toggle group based ones), but regex-based ones have parameters.
- maybe... has a way of explicitly setting priority (in a way that doesn't break existing workflows!)

I think I can see how all this could be done (in some cases I just mean I know whose code I need to study!).

That is not a promise that I'll do it!  If I do it, I'll need to get a couple of volunteers happy to test drive a branch with prototypes...

**Comment by chrisgoringe** [comment]
And hey, kudos for using the Discussions tab for this :)

**Comment by JorgeR81** [comment]
> I'll need to get a couple of volunteers happy to test drive a branch with prototypes...

Psst ... ! @LukeG89 , instructions here ! ... https://github.com/chrisgoringe/cg-use-everywhere/issues/138 ( this branch doesn't exist anymore, but you get the idea )

**Comment by LukeG89** [comment]
@chrisgoringe Basically, as I imagined it, `Any Plus` is a UE? node with steroids.
It starts as a vanilla UE? node with the 3 widgets visible (title, input, group). When a input is added, it adds 3 more widgets to the node (see my first mockup).
Yes, if users add lots of inputs, the UI can get pretty messy... But that's their choice! 😆 


> -  has a right-click option to expose the regex fields, which turns it into a ...? node

About the menu, I was thinking something like that:
After _"Group Restrictions"_ and _"Color Restrictions"_ there is a _"Any Plus options"_ (that can only appear when you right-click on a `Any Plus` node) with the following options:
- Combine title regexes
- Separate title regexes
- Combine group regexes
- Separate group regexes
- Hide input regexes

I don't know if the "Combine/Separate" options can be simplified in just one: For istance, if the node has already all title regexes separated, it only appears the "combine" option, and viceversa. 

For me, the "Hide input regexes" can be optional, you can just collapse the node as usual. But if we want to add an hiding option, I would include all widgets and change it to:

- Hide all regexes

This way, the node will look similar to the `Anything Everywhere3`.

> That is not a promise that I'll do it! If I do it, I'll need to get a couple of volunteers happy to test drive a branch with prototypes...

I like beta testing stuff! If you want, I'm in! 😄

**Comment by chrisgoringe** [comment]
I guess my thought about that description is this: if I fixed the issue with UE node inputs as group node inputs, wouldn't what you are describing be almost as easy to do with a group node consisting of N `UE?` nodes?

I mean, the node described would have a few advantages, but enough?

**Comment by JorgeR81** [comment]
> if I fixed the issue with UE node inputs as group node inputs, wouldn't what you are describing be almost as easy to do with a group node consisting of N UE? nodes?

Yes, I agree with this.
This new node should not be a replacement for grouped UE? nodes.

For me, the advantage of this node ( vs a grouped node ) would be that it's easier and faster to use, in a simpler workflow.

If I wanted to configure something more complex than what @chrisgoringe described, for me it would be easier to use grouped UE? nodes, with the Simple String nodes.
For me, it would be easier to just convert a node group back to nodes, make the internal modifications I want, and then convert back to a group again.
This way, would also help visualise what I want to do, with this more complex configuration. 

![g1](https://github.com/chrisgoringe/cg-use-everywhere/assets/72734038/2df00383-332c-4d68-8c5f-7330bfd6156a)

**Comment by LukeG89** [comment]
@chrisgoringe Yes, you're right: grouping more UE? nodes would be very similar to that. Not as functional, but a good alternative.

> I mean, the node described would have a few advantages, but enough?

I don't know if it's worth it. As I can say, `Any Plus` would be the most flexible node of all UE nodes, the best option for: 
- Having 100% control on all connections;
- Having as many inputs as you need in just one node;
  - Also, adding and removing connections on the fly;
- Making any changes easily. 

Groups are great, but sometimes are also quite annoiyng. For example, if you make any changes (e.g make regexes visible again), the group will reset some changes you made, such as node color, title and size:

https://github.com/chrisgoringe/cg-use-everywhere/assets/161304036/6a4f2425-7a13-41b6-bb6a-5d3f49d73e20


> This new node should not be a replacement for grouped UE? nodes.

@JorgeR81 That's exact the main purpose of this node! Why merging when you can easily do that with `Any Plus`?

> For me, it would be easier to just convert a node group back to nodes, make the internal modifications, and then convert back to a group again.

Easy peasy lemon squeezy! 😆

**Comment by JorgeR81** [comment]
> Groups are great, but sometimes are also quite annoying. For example, if you make any changes (e.g make regexes visible again), the group will reset some changes you made, such as node color, title and size:

Yes, I totally agree. 
But Comfy UI should fix this, eventually ...

Also, I don't need to use groups in every workflow. 

I actually like @chrisgoringe concept, more than our previous concept.

For me, this would still be useful, like 90% of the time.

Most of the time, I just want to send a bunch of UE? nodes to the same [group_regex], or [title _regex].
And I just leave the [input_regex] blank. 

Maybe we should just wait to see what @chrisgoringe comes up with.
See how we can use it, and then see how we can build up from that. 

Even in terms of development and testing, in a more general way, it would be better to start with something simpler.
And then add more features in the future, if needed.

**Comment by LukeG89** [comment]
Another little side effect of using groups is that Autocomplete doesn't work :(
![no_autocomplete](https://github.com/chrisgoringe/cg-use-everywhere/assets/161304036/0b454db3-a91f-471d-891e-7625c57e108c)
____


However guys, `Any Plus` is just an idea I had for creating a node that can satisfy all possible needs.

Honestly, now that you can merge UE? nodes (and it's also optimized) I can perfectly work with that!

Thanks @chrisgoringe for all the hard work! Really appreciated! 😄

**Comment by JorgeR81** [comment]
@chrisgoringe if you have the time to work on your concept, it would still be really useful for me, in most of my use cases, and for users who don't want to use group nodes. 

Even if the `Any Plus` node also does not have autocomplete, I would still be happy with that. 

It's better to type once, in the `Any Plus` node, than to use autocomplete 4 or 5 times, in multiple UE? nodes.

We can also use the Simple String nodes, connected to multiple UE? nodes, but it's more cumbersome to set up, and it takes more space.  
The `Any Plus` node would be a much more elegant solution.
